mainnet_contract/contracts/LiteCheckpoint.sol

```diff
@@ -0,0 +1,424 @@
+        }
+        if (current.length > 0 && next.length > 0) {
+            revert("Malformed Block blockhash");
+        }
+        checkSig(validationParams);
+        if (current.length > 0) {
+            if (
+                uint64(uint256(validationParams.number)) % INIT_EPOCH == 0 &&
+                uint64(uint256(validationParams.number)) / INIT_EPOCH ==
+                epochNum + 1
+            ) {
+                int256 gapNumber = validationParams.number -
+                    (validationParams.number % int256(uint256(INIT_EPOCH))) -
+                    int256(uint256(INIT_GAP));
+                // Edge case at the beginning
+                if (gapNumber < 0) {
+                    gapNumber = 0;
+                }
+                unchecked {
+                    epochNum++;
+                    gapNumber++;
+                }
+
+                if (validators[gapNumber].threshold > 0) {
+                    if (validators[gapNumber].set.length != current.length) {
+                        revert("Mismatched Validators");
+                    }
+                    setLookup(validators[gapNumber].set);
+                    currentValidators = validators[gapNumber];
+                    latestEpoch = blockHash;
+                    currentTree.push(blockHash);
+                } else {
+                    revert("Missing Current Validators");
+                }
+            } else {
+                revert("Invalid Current Block");
+            }
+        }
+
+        if (next.length > 0) {
+            if (
+                uint64(
+                    uint256(
+                        validationParams.number % int256(uint256(INIT_EPOCH))
+                    )
+                ) ==
+                INIT_EPOCH - INIT_GAP + 1 &&
+                uint64(uint256(validationParams.number)) / INIT_EPOCH ==
+                epochNum
+            ) {
+                (bool isValidatorUnique, ) = checkUniqueness(next);
+                if (!isValidatorUnique) revert("Repeated Validator");
+
+                validators[validationParams.number] = Validators({
+                    set: next,
+                    threshold: int256((next.length * 2 * 100) / 3)
+                });
+                latestEpoch = blockHash;
+                currentTree.push(blockHash);
+            } else {
+                revert("Invalid Next Block");
+            }
+        }
+
+        uint256 epochInfo = (uint256(validationParams.number) << 128) |
+            (uint256(validationParams.roundNumber) << 64) |
+            uint256(uint64(int64(-1)));
+        unCommittedLastHash[blockHash] = blockHash;
+        unCommittedTree[blockHash] =
+            (0 << 128) |
+            (uint256(validationParams.roundNumber) << 64) |
+            uint256(validationParams.number);
+
+        headerTree[blockHash] = epochInfo;
+        heightTree[uint64(epochInfo >> 128)] = blockHash;
+        emit SubnetEpochBlockAccepted(blockHash, uint64(epochInfo >> 128));
```

Comment:
please make different emit for block at EPOCH (900) and EPOCH-GAP+1 (451)

Reply:
> please make different emit for block at EPOCH (900) and EPOCH-GAP+1 (451)

i dont know why ,could you pls explain the necessity

becuase in fullCheckpoint , use a emit receive gap/epoch 

same liteCheckpoint

Reply:
Oh, if `uint64(epochInfo >> 128)` can show 900,1800... and 451, 1351... then it is okay. Please confirm what I write here, thanks @GalaxySciTech

Reply:
right